### PR TITLE
Apple pay component advanced and general component alignment

### PR DESCRIPTION
**Pull request by Robert-SD**

### CODE REVIEW

File: android/src/main/kotlin/com/adyen/checkout/flutter/PlatformApi.kt
Comment (Robert-SD) [inline android/src/main/kotlin/com/adyen/checkout/flutter/PlatformApi.kt:1]: Generated code - no review required.

File: ios/Classes/PlatformApi.swift
Comment (Robert-SD) [inline ios/Classes/PlatformApi.swift:1]: Generated code - no review required.

File: lib/src/generated/platform_api.g.dart
Comment (Robert-SD) [inline lib/src/generated/platform_api.g.dart:1]: Generated code - no review required.

File: ios/Classes/components/ComponentSessionFlowDelegate.swift [hunk 1]
```diff
@@ -15,19 +15,20 @@ class ComponentSessionFlowDelegate: AdyenSessionDelegate {
...
-                componentCommunicationModel: componentCommunicationModel,
+            self?.componentFlutterApi.onComponentCommunication(
+                componentCommunicationModel: ComponentCommunicationModel(
+                    type: ComponentCommunicationType.result,
+                    componentId: self?.componentId ?? "",
+                    paymentResult: PaymentResultDTO(
+                        type: PaymentResultEnum.finished,
+                        result: paymentResult
+                    )
+                ),
```
Comment (descorp) [inline ios/Classes/components/ComponentSessionFlowDelegate.swift:31]: I liked `componentCommunicationModel ` as a standalone variable :)

File: ios/Classes/components/applePay/BaseApplePayComponent.swift [hunk 2]
```diff
@@ -7,6 +7,17 @@ class BaseApplePayComponent {
         preconditionFailure("This method must be overridden")
     }
     
+    func handlePaymentEvent(paymentEventDTO: PaymentEventDTO) {}
```
Comment (descorp) [inline ios/Classes/components/applePay/BaseApplePayComponent.swift:10]: Empty method?

File: ios/Classes/components/applePay/BaseApplePayComponent.swift [hunk 3]
```diff
@@ -7,6 +7,17 @@ class BaseApplePayComponent {
...
+    
+    func finalizeAndDismissComponent(success: Bool, completion: @escaping (() -> Void)) {
+        applePayComponent?.finalizeIfNeeded(with: success) { [weak self] in
+            completion()
```
Comment (atmamont) [inline ios/Classes/components/applePay/BaseApplePayComponent.swift:14]: It feels calling completion before the top view controller is dismissed could lead to unecpected behavior if the completion closure involves any navigation changes. I'd suggest to think moving the call inside the second nested completion block (when the view controller gets dismissed)

**Review by atmamont — APPROVED** [review]

File: ios/Classes/components/applePay/BaseApplePayComponent.swift [hunk 4]
```diff
@@ -7,6 +7,17 @@ class BaseApplePayComponent {
...
+            self?.getViewController()?.dismiss(animated: true, completion: { [weak self] in
+                self?.applePayComponent = nil
+            })
```
Comment (descorp) [inline ios/Classes/components/applePay/BaseApplePayComponent.swift:17]:
```suggestion
        applePayComponent?.finalizeIfNeeded(with: success) { [weak self] in
            if let viewController = self?.getViewController() {
                viewController.dismiss(animated: true, completion: { [weak self] in
                self?.applePayComponent = nil
                completion()
            })
            } else {
               completion()
            }
       }
```

File: android/src/main/kotlin/com/adyen/checkout/flutter/components/base/ComponentSessionCallback.kt [hunk 5]
```diff
@@ -41,7 +34,32 @@ abstract class ComponentSessionCallback<T : PaymentComponentState<*>>(
...
                 type = ComponentCommunicationType.RESULT,
                 componentId = componentId,
-                paymentResult = paymentResult
+                paymentResult =
+                    PaymentResultDTO(
+                        type = PaymentResultEnum.FINISHED,
+                        result = paymentResult
+                    )
+            )
+        componentFlutterApi.onComponentCommunication(model) {}
```
Comment (jreij) [inline android/src/main/kotlin/com/adyen/checkout/flutter/components/base/ComponentSessionCallback.kt:43]: Is this empty lambda `{}` intended?

File: android/src/main/kotlin/com/adyen/checkout/flutter/utils/Constants.kt [hunk 6]
```diff
@@ -9,5 +9,6 @@ class Constants {
         const val GOOGLE_PAY_ADVANCED_COMPONENT_KEY = "GOOGLE_PAY_ADVANCED_COMPONENT"
         const val GOOGLE_PAY_SESSION_REQUEST_CODE = 486350
         const val GOOGLE_PAY_ADVANCED_REQUEST_CODE = 486351
+        const val SDK_PAYMENT_CANCELED_IDENTIFIER = "Payment canceled"
```
Comment (jreij) [inline android/src/main/kotlin/com/adyen/checkout/flutter/utils/Constants.kt:12]: If this is not a string that will be displayed or logged anywhere, consider changing to to something that looks like a constant, for example "PAYMENT_CANCELED"

**Review by jreij — DISMISSED** [review]

File: ios/Classes/components/applePay/BaseApplePayComponent.swift (hunk 4 above)
Comment (Robert-SD) [reply ios/Classes/components/applePay/BaseApplePayComponent.swift]: Applied

File: android/src/main/kotlin/com/adyen/checkout/flutter/components/base/ComponentSessionCallback.kt (hunk 5 above)
Comment (Robert-SD) [reply android/src/main/kotlin/com/adyen/checkout/flutter/components/base/ComponentSessionCallback.kt:43]: This lambda is the callback that pigeon uses for a back communication from the Flutter layer. The callback is being generated automatically and not optional. Therefore the empty lambda exists.

File: android/src/main/kotlin/com/adyen/checkout/flutter/utils/Constants.kt (hunk 6 above)
Comment (Robert-SD) [reply android/src/main/kotlin/com/adyen/checkout/flutter/utils/Constants.kt:12]: This string is only used for identifying if the Google Pay sheet is closed by the user. The native SDK provides a ComponentException with "Payment canceled" as the error message. This is used to map the ComponentException to a "PaymentCancelledByUser" outcome. 

This is where the ComponentException gets provided in the native SDK: https://github.com/Adyen/adyen-android/blob/52e05ce148ac454683a5fa92d20900d8e45c9e96/googlepay/src/main/java/com/adyen/checkout/googlepay/internal/ui/DefaultGooglePayDelegate.kt#L156

**Review by descorp — APPROVED** [review]

**Review by atmamont — APPROVED** [review]

**Pull request merged**